python/lsst/daf/base/propertyContainer/propertyContainerContinued.py

```diff
@@ -69,6 +70,40 @@ def _propertyContainerGet(container, name, asArray=False):
     raise lsst.pex.exceptions.TypeError('Unknown PropertySet value type for ' + name)
 
 
+def _propertyContainerGetArray(container, name, asArray=False):
+    """Get a value of unknown type as an array
+
+    Throw TypeError if the item type is not numeric or string
+    (e.g. is PropertyList, PropertySet or PersistablePtr).
+    """
+    if not container.exists(name):
+        raise lsst.pex.exceptions.NotFoundError(name + " not found")
+
+    elemType = _propertyContainerElementTypeName(container, name)
+    if elemType:
+        return getattr(container, "getArray" + elemType)(name)
+
+    raise TypeError("Item {} is not numeric or string".format(name))
+
+
+def _propertyContainerGetScalar(container, name, asArray=False):
+    """Get a value of unknown type as a scalar
+
+    If there are multiple values, return the last value
+
+    Throw TypeError if the item type is not numeric or string
+    (e.g. is PropertyList, PropertySet or PersistablePtr).
+    """
+    if not container.exists(name):
+        raise lsst.pex.exceptions.NotFoundError(name + " not found")
+
+    elemType = _propertyContainerElementTypeName(container, name)
+    if elemType:
+        return getattr(container, "getArray" + elemType)(name)[-1]
+
+    raise TypeError("Item {} is not numeric or string".format(name))
```

Comment:
Do you think that just `getScalar` should be able to return a PropertyList, PropertySet, or PersistablePtr, or that `getArray` should be able to do so as well?

I can see arguments for all 4 options:

- neither (my initial choice): PropertySet and the like are neither scalars nor arrays, so using `get` makes for cleaner and less surprising code. Also the user is less likely to be surprised by retrieving a `PropertySet` when they expected a number, string or list of either of those.
- both: it's simple to return the obvious object if it's not numeric or string-like. It allows the user to iterate over all names and gets something reasonable back. The user just has to be careful with types, but that's implied by a container such as this anyway. Also this allows us to (in theory) deprecate `get`.
- getArray only: getArray returns the most "native" format -- since internally numeric and string data is stored in an array. But that's arguably an internal detail, so I consider that a very weak argument.
- getScalar only: getArray always returns an array and PropertySet and the like are not arrays, they are objects, so they are more like scalars than arrays.

I'll see about sharing code. I'm personally inclined to get rid of the `asArray` argument to `get` but that may be out of scope.

Reply:
After talking to @ktlim we agreed that `getScalar` will be able to return `PropertySet` and similar items, but that `getArray` will continue to raise. If I am successful in removing the `asArray` argument of `get` (without breaking existing code) then this will result in a fairly logical API. With the `asArray` argument it's a bit weird because `get(name, asArray=True)` can return `PropertySet` and `getArray` cannot.

Reply:
As expressed in Slack, I'm OK with either "both" or "getScalar only".